### ISSUE TITLE
When self-edges (subclass_of) exist, the results / edges are odd

### ISSUE THREAD

**Issue by colleenXu** [biothings/biothings_explorer#734]
I've found a case where all of BTE's results use subclassing and aux/support-graphs, but the "subclass_of" edge involved is a self-edge. Which means BTE actually has direct edges from the starting ID, but it doesn't look that way in the response. 

<details><summary>click here for screenshot examples</summary>
<p>

![Screen Shot 2023-10-02 at 4 58 09 PM](https://github.com/biothings/biothings_explorer/assets/43731687/589ca67e-d6e9-475d-9f8a-cb1c27916fde)

![Screen Shot 2023-10-02 at 4 58 17 PM](https://github.com/biothings/biothings_explorer/assets/43731687/a1dc7c15-e5a6-416a-a7d2-6284620296f8)

</p>
</details> 

This seems...not ideal. I'd expect no self-subclass edge and no nested-edges/support-graphs in this situation. 

---

<details><summary>How to recreate</summary>
<p>

query only suppKG through BTE (`http://localhost:3000/v1/smartapi/b48c34df08d16311e3bca06b135b828d/query`)

The query uses a UMLS ID for ulcerative colitis

```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "ids": ["UMLS:C0009324"],
                    "categories": ["biolink:Disease"]
                },
                "n1": {
                    "categories": ["biolink:SmallMolecule"]
                }
            },
            "edges": {
                "e01": {
                    "subject": "n0",
                    "object": "n1"
                }
            }
        }
    }
}
```

</p>
</details> 

Response JSON: [subclass_suppKG.txt](https://github.com/biothings/biothings_explorer/files/12787172/subclass_suppKG.txt)

In the response, you can see:
* all 57 results have edge-binding IDs that end in "-via-subclass" (aka these are nested edges). This can be seen in the ARAX-UI (where clicking on the nested edges gets you to the support graph)
* there's only 1 edge with the predicate `biolink:subclass_of`, which appears to be `expanded-MONDO:0005101-subclass_of-MONDO:0005101`

Note: uncovered in the course of testing https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1744327615 (which involves a custom-modified vresion of the BioThings SuppKG yaml), but this behavior happens even when this behavior isn't being tested (aka you use the suppKG info from the SmartAPI registry, as mentioned in "how to recreate"). 

---

My investigation so far:

<details><summary>(1) Strangely, this doesn't seem to happen with other UMLS IDs or queries (that all involve ID/node expansion), which makes me wonder if the bug originates in the node-expansion module and is limited to only specific IDs...</summary>
<p>

* Diverticulosis Disease UMLS:C1510475 -> SmallMolecule: all results involve direct edges. seems fine (no self-edges). 
* muscular dystrophy Disease UMLS:C0026850 -> SmallMolecule: all results involve subclassing/nested-edges, but no self-edges so it looks okay...
* reverse: Fluconazole (SmallMolecule `UMLS:C0016277`) -> Disease: all results involve direct edges

</p>
</details> 

<details><summary>(2) I double-checked whether NodeNorm of UMLS IDs might be involved (maybe this is happening when there's subclass IDs involved, but NodeNorm merges them into being the same entity?). But this doesn't seem to be the case...</summary>
<p>

* I can see 2 UMLS IDs in the node-normalized Disease Node: C0009324 (starting ID) and C0375359 (potential subclass?)
* But [suppKG doesn't have any data using C0375359](https://pending.biothings.io/suppkg/query?q=subject.umls:C0375359%20OR%20object.umls:C0375359)
* So all of the data for this response should be coming from [C0009324](https://pending.biothings.io/suppkg/query?q=object.umls:C0009324%20AND%20predicate:TREATS)

</p>

**Comment by colleenXu** [comment]
Note: uncovered in the course of testing https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1744327615, but this behavior happens even when this behavior isn't being tested (aka you use the suppKG info from the SmartAPI registry).

**Comment by andrewsu** [comment]
best guess is that this problem stems from the presence of this edge:

```
                "expanded-MONDO:0005101-subclass_of-MONDO:0005101": {
                    "predicate": "biolink:subclass_of",
                    "subject": "MONDO:0005101",
                    "object": "MONDO:0005101",
                    "attributes": [],
                    "sources": [
                        {
                            "resource_id": "infores:umls",
                            "resource_role": "primary_knowledge_source"
                        },
                        {
                            "resource_id": "infores:service-provider-trapi",
                            "resource_role": "aggregator_knowledge_source"
                        }
                    ]
                },
```
solution / next debugging: filter out in node expansion module?

**Comment by tokebe** [comment]
Confirmed, node expansion `getDescendants()` is not supposed to return input ids, but is (that's the only thing that could cause this behavior). So, something in the current ontology files is either wrong, or using self-subclasses for specific reasons that we don't care for.

Easiest fix: run a check at the end of `getDescendants()` and remove any ids that were present in the input.

More complicated fix: If these self-subclasses are errors (and not intended) in the ontologies, we could look into updating our ontology files, or adding a cron for routinely updating these ontology files.

**Comment by andrewsu** [comment]
The issue linked above shows another self-edge using the `has_phenotype` predicate, and I'd argue that this edge is also not useful.  If we wanted to remove these, I assume the `getDescendants()` fix would not be sufficient here?

NOTE that these are not errors in the source files per se, but just edges that BTE should screen out as non-informative...

**Comment by tokebe** [comment]
Yes, this kind of edge has to be coming from an upstream source rather than node expansion. We'd probably have to filter these out prior to or during results assembly.

**Comment by colleenXu** [comment]
From my digging, this is coming from the "Disease-PhenotypicFeature conflation in Translator (specifically NodeNorm)". I suspect the original data for the "self-edge" is `Disease` `OMIM:600807` susceptibility to asthma  <-> `PhenotypicFeature` `HP:0002099` asthma 

I'm not sure if that complicates "filtering out self-edges" or points to an issue with NodeNorm...

---

What I've found:

<details><summary>In the ARAX UI, we can see that the self-edge's source is MyDisease's HPO annotations</summary>
<p>

[Link](https://arax.ci.transltr.io/?r=3a84530f-e142-48b3-8da2-1be4c99d8488), ctrl-F for result 252

![Screen Shot 2023-10-10 at 12 23 55 PM](https://github.com/biothings/biothings_explorer/assets/43731687/dd8c330f-a8b7-4c92-8573-c9c331314164)

</p>
</details> 

<details><summary>However, the data in MyDisease's HPO annotations isn't a self-edge</summary>
<p>

The MyDisease's x-bte-operations for HPO annotations are: 
* **OMIM Disease <-> HP PhenotypicFeature**
* ORPHANET Disease <-> HP PhenotypicFeature

There are [<100 diseases in the HPO annotations with asthma (HP:0002099) as a phenotype](https://mydisease.info/v1/query?q=hpo.phenotype_related_to_disease.hpo_id:%22HP:0002099%22&fields=hpo,size=100). 

[One of them](https://mydisease.info/v1/query?q=hpo.phenotype_related_to_disease.hpo_id:%22HP:0002099%22%20AND%20hpo.disease_name:asthma&fields=hpo,size=100) is `OMIM:600807` ["Asthma, susceptibility to"](https://omim.org/entry/600807). 

We can also see this info on the [HPO website](https://hpo.jax.org/app/browse/term/HP:0002099) (after ctrl-f the disease associations section)

</p>
</details> 

<details><summary>Looking back at ARAX UI, we can see that NodeNorm considers the OMIM and HP IDs equivalent to each other and to the Disease asthma</summary>
<p>

See the blue arrows

Both the "susceptibility to asthma" OMIM:600807 and phenotype asthma "HP:0002099" show up in the Disease asthma's equivalent IDs. 

![Screen Shot 2023-10-10 at 12 24 25 PM](https://github.com/biothings/biothings_explorer/assets/43731687/28b71999-38f1-42ef-be85-360818c80344)

</p>
</details>

**Comment by tokebe** [comment]
From discussion with @colleenXu: This could be a check prior after record retrieval and node normalization, where we remove any record with the same input and output. There _should_ be no loss of information any case like this would have a direct-edge result from the first template. @andrewsu Should we go ahead with this fix?

**Comment by andrewsu** [comment]
👍I support this fix

**Comment by colleenXu** [comment]
@tokebe 

Can we clarify your [earlier post](https://github.com/biothings/biothings_explorer/issues/734#issuecomment-1758289279)? I found the wording a little confusing...

> This could be a check prior after record retrieval and node normalization, where we remove any record with the same input and output.

* What does "prior" mean here? Is this a typo?
* For "record with the same same input and output", I'm thinking this means the "records where, after node normalization, the input and output nodes have the same primaryCurie - meaning the record now represents a self-edge". Do you agree with that?
* And this should be happening after every QEdge execution, so "self-edge"-related nodes aren't used in the next QEdge's execution. Do you agree with that?

**Comment by colleenXu** [comment]
And on:

> There should be no loss of information any case like this would have a direct-edge result from the first template.

There's a nebulous concern that we'll lose information by removing these self-edges. While Jackson and I discussed it and reached consensus that it's probably a theoretical edge case that's not really something to worry about, I want to write out my thinking...

(1) We imagined two possible scenarios where creative-mode template QEdges may inadvertently retrieve records that become self-edges: 
* DiseaseOrPheno X -(has_phenotype)-> itself
* Gene Y -(regulates / interacts_with)-> itself

<details><summary>(2) preventing loss of info: templateGroups + NodeNorm</summary>
<p>

For paths w/ self-edges right now, we should be able to retrieve the equivalent path w/o the self-edge because:
* the templateGroups are set up to run the direct/shorter QGraphs first
* NodeNorm is giving us the set of equivalent IDs, which we'll use when running those direct/shorter QGraphs

</p>
</details> 

---

And new (not discussed earlier w/ @tokebe Jackson...)

<details><summary>(3) Actually, this could theoretically remove some "opposite of" problems?</summary>
<p>

For example, with Chem-increases-Gene, we have a template that's "Chem-DecreaseAnotherGeneThatDownregs-Gene". 

Theoretically, an "opposite-of-what-I-asked-for" issue can happen if the "Gene-downregs-Gene" edge is a self-edge: 
"Chem A -(deceases)-> Gene Y, and Gene Y downregs itself" == "Chem A -(decreases)-> Gene Y"

By removing the self-edge "Gene Y downregs itself", we should prevent this situation from happening...

@tokebe In this scenario, will BTE automatically remove the "Chem A -(decreases)-> Gene Y" edge if it's not part of any complete path? 

</p>
</details> 

(4) in non-creative mode, we will be losing info: removing valid paths for longer QGraphs. However, I think this may be fine, because it feels kinda odd to depend on self-edges to return "shorter path" info for longer QGraphs...
